### PR TITLE
[DEN-37] BETA Route DNS Implementation

### PR DESCRIPTION
**Pull request by wdawg33**
NOTE: Route DNS support uses beta APIs, therefore feature support in the SDK should be considered beta status.

1. Implemented BETA Route DNS feature support (Master Server Group, Primary Zone, Secondary Zone Group, TSIG, Load Balancing and Failover Group)
2. Created example file for using Route DNS via the SDK

### CODE REVIEW

File: edgecast/routedns/group.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+	resp, err := svc.Client.SendRequestWithStringResponse(request)
+
+	if err != nil {
+		return nil, fmt.Errorf("AddGroup->API Response Error: %v", err)
+	}
+
+	groupID, err := strconv.Atoi(*resp)
+	if err != nil {
+		return nil, fmt.Errorf(
```
Comment (stevenpaz) [inline edgecast/routedns/group.go:62]: Might be helpful to also print out the string text that failed to parse (if the error does not already include it)

File: edgecast/routedns/group.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	resp, err := svc.Client.SendRequestWithStringResponse(request)
+
+	if err != nil {
+		return fmt.Errorf("UpdateGroup->API Response Error: %v", err)
+	}
+
+	// Group ID changes when updating a group. Update Group object with latest
+	// ID
+	groupID, err := strconv.Atoi(*resp)
+	if err != nil {
```
Comment (stevenpaz) [inline edgecast/routedns/group.go:90]: similar to above

File: edgecast/routedns/masterservergroup.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+	parsedResponse := []*MasterServerGroupAddGetOK{}
+
+	_, err = svc.Client.SendRequest(request, &parsedResponse)
+	if err != nil {
+		return nil, fmt.Errorf("GetMasterServerGroup: %v", err)
+	}
+
+	// Single object get always returns an array of one
+	return parsedResponse[0], nil
```
Comment (stevenpaz) [inline edgecast/routedns/masterservergroup.go:60]: May be good to check if len() > 0, then maybe return nil if that is the case

File: edgecast/routedns/masterservergroup.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+	if err != nil {
+		return nil, fmt.Errorf(
+			"AddMasterServerGroup->API Response Error: %v",
+			err,
+		)
+	}
+
+	// Add operation always returns an array of one
+	return parsedResponse[0], nil
```
Comment (stevenpaz) [inline edgecast/routedns/masterservergroup.go:96]: Same as above regarding length check

File: edgecast/routedns/secondaryzonegroup.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+	resp, err := svc.Client.SendRequest(request, &parsedResponse)
+	log.Printf("GetSecondaryZoneGroup:%v", resp)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"GetSecondaryZoneGroup->API Response Error: %v",
+			err,
+		)
+	}
+
+	return parsedResponse[0], nil
```
Comment (stevenpaz) [inline edgecast/routedns/secondaryzonegroup.go:41]: Same as above regarding length check

File: edgecast/routedns/tsig.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+	}
+
+	resp, err := svc.Client.SendRequestWithStringResponse(request)
+
+	if err != nil {
+		return nil, fmt.Errorf("AddTsig->API Response Error: %v", err)
+	}
+
+	tsigID, err := strconv.Atoi(*resp)
+	if err != nil {
```
Comment (stevenpaz) [inline edgecast/routedns/tsig.go:53]: Same as above regarding printing the string if it failed to parse as an int

File: edgecast/routedns/zone.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+	if err != nil {
+		return nil, fmt.Errorf(
+			"AddZone->->API Response Error: %v",
+			err,
+		)
+	}
+
+	// Success response body contains only the zone ID
+	zoneID, err := strconv.Atoi(*resp)
```
Comment (stevenpaz) [inline edgecast/routedns/zone.go:60]: Same as above regarding printing the string if it failed to parse as an int

File: edgecast/routedns/group.go (hunk 1 above)
Comment (wdawg33) [reply edgecast/routedns/group.go:62]: Tested and confirmed the error does include the string.

e.g. resp, err := strconv.Atoi("notAnum")
result: parsing "notAnum": invalid syntax

File: edgecast/routedns/group.go (hunk 2 above)
Comment (wdawg33) [reply edgecast/routedns/group.go:90]: Confirmed error contains passed string

File: edgecast/routedns/tsig.go (hunk 6 above)
Comment (wdawg33) [reply edgecast/routedns/tsig.go:53]: Confirmed error contains passed string

File: edgecast/routedns/zone.go (hunk 7 above)
Comment (wdawg33) [reply edgecast/routedns/zone.go:60]: Confirmed error contains passed string

File: edgecast/routedns/masterservergroup.go (hunk 4 above)
Comment (wdawg33) [reply edgecast/routedns/masterservergroup.go:110]: Good call, added check!

**Review by stevenpaz — APPROVED** [review]

**Pull request merged**